components/panic_hook/Cargo.toml

```diff
@@ -0,0 +1,7 @@
+[package]
+name = "panic_hook"
+version = "0.0.1"
+publish = false
+
+[lib]
```

Comment:
by default it is `path = "src/lib.rs"`

Reply:
Why are `codec` and `panic_hook` in the same directory as `test_*`?

Reply:
because they are all components.

Reply:
To make files not too deep. Otherwise you will only see "Cargo.toml" and "src/" in each component. This pattern can be seen in some most popular Rust projects that using workspace, like [servo](https://github.com/servo/servo/tree/master/components/allocator) and [rust](https://github.com/rust-lang/rust/tree/master/src/liballoc).

Reply:
Also this is aligned with other workspace libraries in `components/`, which are not stayed in `src/`.

Reply:
There are also many projects keep the `src` directory, like `futures` or `tokio`. I think it's better to follow convention and keep it implicit.

> Also this is aligned with other workspace libraries in components/, which are not stayed in src/.

That's not a problem as long as all the projects use the same layout.

Reply:
@BusyJay They are independent crates, for example, futures-core and futures-util, so they need extra files like LICENSE in each root. For us, it is not needed.

Reply:
@BusyJay Even if we are going to change the layout, it is off-topic for this PR.

Reply:
> They are independent crates, for example, futures-core and futures-util, so they need extra files like LICENSE in each root. 

I don't see any relations between independence of crates and project layouts.

>  is off-topic for this PR.

Agree. Let's do it in separate PR.